dbfilter_from_header/override.py

```diff
@@ -6,24 +6,99 @@
 import logging
 import re
 
+import psycopg2
+
+import odoo.service.db
 from odoo import http
+from odoo.http import get_default_session, root
 from odoo.tools import config
 
+_logger = logging.getLogger(__name__)
+
 db_filter_org = http.db_filter
 
 
-def db_filter(dbs, httprequest=None):
-    dbs = db_filter_org(dbs, httprequest)
+def db_filter(dbs, host=None, httprequest=None):
+    dbs = db_filter_org(dbs, host)
     httprequest = httprequest or http.request.httprequest
     db_filter_hdr = httprequest.environ.get("HTTP_X_ODOO_DBFILTER")
     if db_filter_hdr:
         dbs = [db for db in dbs if re.match(db_filter_hdr, db)]
     return dbs
 
 
+# we don't rely on original http.db_list
+# because it calls db_filter(dbs, host)
+# while we also need to pass httprequest
+def db_list(force=False, host=None, httprequest=None):
+    try:
+        dbs = odoo.service.db.list_dbs(force)
+    except psycopg2.OperationalError:
+        return []
+    return db_filter(dbs, host, httprequest)
+
+
+# we need to override Request._get_session_and_dbname
+# to pass self.httprequest to db_filter() and db_list()
+# because when _get_session_and_dbname is called,
+# http.request is not bound yet
+class RequestNew(http.Request):
+    def _get_session_and_dbname(self):
+        # The session is explicit when it comes from the query-string or
+        # the header. It is implicit when it comes from the cookie or
+        # that is does not exist yet. The explicit session should be
+        # used in this request only, it should not be saved on the
+        # response cookie.
+        sid = self.httprequest.args.get("session_id") or self.httprequest.headers.get(
+            "X-Openerp-Session-Id"
+        )
+        if sid:
+            is_explicit = True
+        else:
+            sid = self.httprequest.cookies.get("session_id")
+            is_explicit = False
+
+        if sid is None:
+            session = root.session_store.new()
+        else:
+            session = root.session_store.get(sid)
+            session.sid = sid  # in case the session was not persisted
+        session.is_explicit = is_explicit
+
+        for key, val in get_default_session().items():
+            session.setdefault(key, val)
+        if not session.context.get("lang"):
+            session.context["lang"] = self.default_lang()
+
+        dbname = None
+        host = self.httprequest.environ["HTTP_HOST"]
+        if session.db and db_filter(
+            [session.db], host=host, httprequest=self.httprequest
+        ):
+            dbname = session.db
+        else:
+            all_dbs = db_list(force=True, host=host, httprequest=self.httprequest)
```

Comment:
At the very least, mark this line as diverging from upstream, with the rest being identical. Also at the very least, mark the date and commit from which the upstream code was taken.